File: player/src/app/lib/websocket/protocols/MessageTypes.ts
// Define message types for type-safety
export enum MessageType {
  // Connection Management
  HANDSHAKE = 'HANDSHAKE',
  HANDSHAKE_RESPONSE = 'HANDSHAKE_RESPONSE',
  CONNECTION_KEEP_ALIVE = 'CONNECTION_KEEP_ALIVE',
  
  // Playback Control
  PLAYBACK_CONTROL = 'PLAYBACK_CONTROL',
  PLAYBACK_STATE = 'PLAYBACK_STATE',
  UPDATE_TRACK_CONTEXT = 'UPDATE_TRACK_CONTEXT',
  
  // Voice Communication
  VOICE_STREAM_START = 'VOICE_STREAM_START',
  VOICE_STREAM_END = 'VOICE_STREAM_END',
  VOICE_DATA = 'VOICE_DATA',
  
  // Application State
  STATE_UPDATE = 'STATE_UPDATE',
  STATE_SYNC_REQUEST = 'STATE_SYNC_REQUEST',
  STATE_SYNC_RESPONSE = 'STATE_SYNC_RESPONSE',
  
  // Notifications/UI
  NOTIFICATION = 'NOTIFICATION',
  ERROR = 'ERROR',

  // AI Voice Streaming
  AI_VOICE_STREAMING_START = 'AI_VOICE_STREAMING_START',
  AI_VOICE_STREAMING_END = 'AI_VOICE_STREAMING_END',
}

// Protocol Payload Types
export interface HandshakePayload {
  client: string;
  version: string;
  capabilities: string[];
}

export interface HandshakeResponsePayload {
  sessionId: string;
  features: string[];
  serverTime: number;
}

export interface PlaybackControlPayload {
  action: 'play' | 'pause' | 'seek' | 'next' | 'previous';
  trackId?: string;
  position?: number;
}

export interface PlaybackStatePayload {
  isPlaying: boolean;
  currentTrack?: {
    id: string;
    position: number;
    duration: number;
  };
}

export interface UpdateTrackContextPayload {
  trackId: string;
  duration: number;
  currentTime: number;
}

export interface VoiceStreamStartPayload {
  format: string;
  sampleRate: number;
  channels: number;
}

export interface StateUpdatePayload {
  path: string;
  value: any;
}

export interface NotificationPayload {
  type: 'info' | 'warning' | 'error' | 'success';
  message: string;
  duration?: number;
}

export interface AIVoiceStreamingStartPayload {
  trackId: string;
  duration: number;
  currentTime: number;
}

export interface AIVoiceStreamingEndPayload {
  trackId: string;
  duration: number;
  currentTime: number;
}

// Combine all message types with their payloads
export type ProtocolMessage<T extends MessageType> = 
  T extends MessageType.HANDSHAKE ? { type: T; payload: HandshakePayload } :
  T extends MessageType.HANDSHAKE_RESPONSE ? { type: T; payload: HandshakeResponsePayload } :
  T extends MessageType.PLAYBACK_CONTROL ? { type: T; payload: PlaybackControlPayload } :
  T extends MessageType.PLAYBACK_STATE ? { type: T; payload: PlaybackStatePayload } :
  T extends MessageType.UPDATE_TRACK_CONTEXT ? { type: T; payload: UpdateTrackContextPayload } :
  T extends MessageType.AI_VOICE_STREAMING_START ? { type: T; payload: AIVoiceStreamingStartPayload } :
  T extends MessageType.AI_VOICE_STREAMING_END ? { type: T; payload: AIVoiceStreamingEndPayload } :
  T extends MessageType.VOICE_STREAM_START ? { type: T; payload: VoiceStreamStartPayload } :
  T extends MessageType.NOTIFICATION ? { type: T; payload: NotificationPayload } :
  { type: T; payload: any }; 

// Client message types (used when sending)
export type ClientMessage = 
  | ProtocolMessage<MessageType.HANDSHAKE>
  | ProtocolMessage<MessageType.PLAYBACK_CONTROL>
  | ProtocolMessage<MessageType.UPDATE_TRACK_CONTEXT>
  | ProtocolMessage<MessageType.AI_VOICE_STREAMING_START>
  | ProtocolMessage<MessageType.AI_VOICE_STREAMING_END>
  | ProtocolMessage<MessageType.VOICE_STREAM_START>;

// Server message types (used when receiving)
export type ServerMessage = 
  | ProtocolMessage<MessageType.HANDSHAKE_RESPONSE>
  | ProtocolMessage<MessageType.STATE_UPDATE>
  | ProtocolMessage<MessageType.PLAYBACK_STATE>
  | ProtocolMessage<MessageType.NOTIFICATION>
  | ProtocolMessage<MessageType.ERROR>;

File: backend/src/curiousikid/websocket/message_types.py
from enum import Enum
from typing import Dict, List, Optional, Union, Any
from pydantic import BaseModel


class MessageType(str, Enum):
    # Connection Management
    HANDSHAKE = "HANDSHAKE"
    HANDSHAKE_RESPONSE = "HANDSHAKE_RESPONSE"
    CONNECTION_KEEP_ALIVE = "CONNECTION_KEEP_ALIVE"
    
    # Track Context
    UPDATE_TRACK_CONTEXT = "UPDATE_TRACK_CONTEXT"
    
    # Playback Control
    PLAYBACK_CONTROL = "PLAYBACK_CONTROL"
    PLAYBACK_STATE = "PLAYBACK_STATE"
    
    # Voice Communication
    AI_VOICE_STREAMING_START = "AI_VOICE_STREAMING_START"
    AI_VOICE_STREAMING_END = "AI_VOICE_STREAMING_END"
    VOICE_DATA = "VOICE_DATA"
    
    # Application State
    STATE_UPDATE = "STATE_UPDATE"
    STATE_SYNC_REQUEST = "STATE_SYNC_REQUEST"
    STATE_SYNC_RESPONSE = "STATE_SYNC_RESPONSE"
    
    # Notifications/UI
    NOTIFICATION = "NOTIFICATION"
    ERROR = "ERROR"


# Base message model
class Message(BaseModel):
    type: MessageType
    payload: Dict[str, Any]

class UpdateTrackContextPayload(BaseModel):
    """Payload for track context updates from the client."""
    trackId: str
    duration: float  # Change to float to better match JavaScript's number type
    currentTime: float

# Request payload models
class HandshakePayload(BaseModel):
    user_id: str


class PlaybackControlPayload(BaseModel):
    action: str  # 'play' | 'pause' | 'seek' | 'next' | 'previous'
    trackId: Optional[str] = None
    position: Optional[float] = None


class AIVoiceStreamingStartPayload(BaseModel):
    trackId: str
    duration: float
    currentTime: float

class AIVoiceStreamingEndPayload(BaseModel):
    trackId: str
    duration: float
    currentTime: float

class StateSyncRequestPayload(BaseModel):
    clientTime: int




# Response payload models
class HandshakeResponsePayload(BaseModel):
    sessionId: str


class PlaybackStatePayload(BaseModel):
    isPlaying: bool
    currentTrack: Optional[Dict[str, Any]] = None


class StateUpdatePayload(BaseModel):
    path: str
    value: Any


class NotificationPayload(BaseModel):
    type: str  # 'info' | 'warning' | 'error' | 'success'
    message: str
    duration: Optional[int] = None


class ErrorPayload(BaseModel):
    code: str
    message: str
    details: Optional[Dict[str, Any]] = None


class StateSyncResponsePayload(BaseModel):
    serverTime: int
    state: Dict[str, Any] 